kyo-core/shared/src/main/scala/kyo/Queue.scala

```diff
@@ -13,189 +12,332 @@ import scala.annotation.tailrec
+                    def drain()(using AllowUnsafe)           = underlying.drain()
+                    def close()(using Frame, AllowUnsafe)    = underlying.close()
+                    def closed()(using AllowUnsafe): Boolean = underlying.closed()
+                end new
+            end initSliding
+        end Unsafe
     end Unbounded
 
-    /** Initializes a new queue with the specified capacity and access pattern.
-      *
-      * @param capacity
-      *   the capacity of the queue
-      * @param access
-      *   the access pattern (default is MPMC)
-      * @return
-      *   a new Queue instance
-      */
-    def init[A](capacity: Int, access: Access = Access.MultiProducerMultiConsumer)(using frame: Frame): Queue[A] < IO =
-        IO {
+    /** WARNING: Low-level API meant for integrations, libraries, and performance-sensitive code. See AllowUnsafe for more details. */
+    abstract class Unsafe[A]:
+        def capacity: Int
+        def size()(using AllowUnsafe): Result[Closed, Int]
+        def empty()(using AllowUnsafe): Result[Closed, Boolean]
+        def full()(using AllowUnsafe): Result[Closed, Boolean]
+        def offer(v: A)(using AllowUnsafe): Result[Closed, Boolean]
+        def poll()(using AllowUnsafe): Result[Closed, Maybe[A]]
+        def peek()(using AllowUnsafe): Result[Closed, Maybe[A]]
+        def drain()(using AllowUnsafe): Result[Closed, Seq[A]]
+        def close()(using Frame, AllowUnsafe): Maybe[Seq[A]]
+        def closed()(using AllowUnsafe): Boolean
+        final def safe: Queue[A] = this
+    end Unsafe
+
+    /** WARNING: Low-level API meant for integrations, libraries, and performance-sensitive code. See AllowUnsafe for more details. */
+    object Unsafe:
+
+        abstract private class Closeable[A](initFrame: Frame) extends Unsafe[A]:
+            import AllowUnsafe.embrace.danger
+            final protected val _closed = AtomicRef.Unsafe.init(Maybe.empty[Result.Error[Closed]])
+
+            final def close()(using frame: Frame, allow: AllowUnsafe) =
+                val fail = Result.Fail(Closed("Queue", initFrame, frame))
+                Maybe.when(_closed.cas(Maybe.empty, Maybe(fail)))(_drain())
+            end close
+
+            final def closed()(using AllowUnsafe) = _closed.get().isDefined
+
+            final def drain()(using AllowUnsafe): Result[Closed, Seq[A]] = op(_drain())
+
+            protected def _drain(): Seq[A]
+
+            protected inline def op[A](inline f: => A): Result[Closed, A] =
+                _closed.get().getOrElse(Result(f))
+
+            protected inline def offerOp[A](inline f: => Boolean, inline raceRepair: => Boolean): Result[Closed, Boolean] =
+                _closed.get().getOrElse {
+                    val result = f
+                    if result && _closed.get().isDefined then
+                        Result(raceRepair)
+                    else
+                        Result(result)
+                    end if
+                }
```

Comment:
This is part of the fix for the other bug I identified. There's a race between the queue getting closed and a new item being added concurrently. The thread that is adding the new item might see the queue still not closed but it gets closed before the item is added to the underlying queue. To resolve the race, an extra check is performed after the operation and a repair is performed.